lxd/deployments/deployment.go

```diff
@@ -0,0 +1,308 @@
+		return tx.RenameDeployment(ctx, d.id, newName)
+	})
+	if err != nil {
+		return err
+	}
+
+	// Apply changes internally.
+	d.info.Name = newName
+	return nil
+}
+
+// Delete the deployment.
+func (d *Deployment) Delete() error {
+	usedBy, err := d.UsedBy()
+	if err != nil {
+		return err
+	}
+
+	if len(usedBy) > 0 {
+		return fmt.Errorf("Cannot delete deployment, it is in use by %v", usedBy)
+	}
+
+	return d.state.DB.Cluster.Transaction(d.state.ShutdownCtx, func(ctx context.Context, tx *db.ClusterTx) error {
+		return tx.DeleteDeployment(ctx, d.id)
+	})
+}
+
+// DeploymentShapeGet returns a deployment shape.
+func (d *Deployment) DeploymentShapeGet(deploymentShapeName string) (deploymentShape *api.DeploymentShape, err error) {
+	err = d.state.DB.Cluster.Transaction(d.state.ShutdownCtx, func(ctx context.Context, tx *db.ClusterTx) error {
+		dbDeploymentShape, err := tx.GetDeploymentShape(ctx, d.id, deploymentShapeName)
+		if err != nil {
+			return err
+		}
+
+		deploymentShape = &dbDeploymentShape.DeploymentShape
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return deploymentShape, nil
+}
+
+// DeploymentShapeCreate creates a new instance set.
+func (d *Deployment) DeploymentShapeCreate(req api.DeploymentShapesPost) error {
+	// tomp TODO validation.
+
+	return d.state.DB.Cluster.Transaction(d.state.ShutdownCtx, func(ctx context.Context, tx *db.ClusterTx) error {
+		_, err := tx.CreateDeploymentShape(ctx, d.id, req)
+		return err
+	})
+}
+
+// DeploymentShapeDelete deletes a deployment shape.
+func (d *Deployment) DeploymentShapeDelete(deploymentShapeName string) error {
+	return d.state.DB.Cluster.Transaction(d.state.ShutdownCtx, func(ctx context.Context, tx *db.ClusterTx) error {
+		deploymentShape, err := tx.GetDeploymentShape(ctx, d.id, deploymentShapeName)
+		if err != nil {
+			return err
+		}
+
+		return tx.DeleteDeploymentShape(ctx, d.id, deploymentShape.ID)
+	})
+}
+
+// DeploymentShapeUpdate updates a deployment shape.
+func (d *Deployment) DeploymentShapeUpdate(deploymentShapeName string, req api.DeploymentShapePut) error {
+	return d.state.DB.Cluster.Transaction(d.state.ShutdownCtx, func(ctx context.Context, tx *db.ClusterTx) error {
+		deploymentShape, err := tx.GetDeploymentShape(ctx, d.id, deploymentShapeName)
+		if err != nil {
+			return err
+		}
+
+		return tx.UpdateDeploymentShape(ctx, d.id, deploymentShape.ID, req)
+	})
+}
+
+// DeploymentShapeRename renames a deployment shape.
+func (d *Deployment) DeploymentShapeRename(deploymentShapeName string, req api.DeploymentShapePost) error {
+	return d.state.DB.Cluster.Transaction(d.state.ShutdownCtx, func(ctx context.Context, tx *db.ClusterTx) error {
+		deploymentShape, err := tx.GetDeploymentShape(ctx, d.id, deploymentShapeName)
+		if err != nil {
+			return err
+		}
+
+		return tx.RenameDeploymentShape(ctx, d.id, deploymentShape.ID, req.Name)
+	})
+}
+
+// DeploymentKey represents a deployment key for a deployment.
+type DeploymentKey struct {
```

Comment:
I wonder if we should just have the deployment key functions attached to the deployment type itself, and accepting a shape name/id argument instead. This would avoid introduce another type definition for each concept (the third in this PR).

Reply:
This would be similar to what we've done in networking with Forwards for instance https://github.com/canonical/lxd/blob/main/lxd/network/driver_ovn.go#L4594